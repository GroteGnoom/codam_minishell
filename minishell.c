/* ************************************************************************** */
/*                                                                            */
/*                                                        ::::::::            */
/*   minishell.c                                        :+:    :+:            */
/*                                                     +:+                    */
/*   By: sde-rijk <sde-rijk@student.codam.nl>         +#+                     */
/*                                                   +#+                      */
/*   Created: 2021/12/13 10:16:10 by sde-rijk      #+#    #+#                 */
/*   Updated: 2022/01/04 14:08:11 by sde-rijk      ########   odam.nl         */
/*                                                                            */
/* ************************************************************************** */

#include <stdio.h>
#include <readline/readline.h>
#include <readline/history.h>
#include <stdlib.h>
#include "Libft/libft.h"
#include "minishell.h"
#include "get_next_line/get_next_line.h"

#define PROMPT "> "

int	check_for_redirections(int *last_exit_status, t_part *parts, \
t_env *s_env, int nr_parts)
{
	int	exec;
	int	i;

	i = 0;
	exec = 0;
	while (i < nr_parts)
	{
		if (parts[i].type == SPECIAL)
		{
			if (!ft_strcmp(parts[i].part, "<"))
				*last_exit_status = redirect_in(nr_parts, parts, s_env, &exec);
			if (!ft_strcmp(parts[i].part, ">"))
				*last_exit_status = redirect_out(nr_parts, parts, s_env, &exec);
			if (!ft_strcmp(parts[i].part, ">>"))
				*last_exit_status = redirect_out_app(nr_parts, parts, \
				s_env, &exec);
			if (!ft_strcmp(parts[i].part, "<<"))
				*last_exit_status = redirect_here_doc(nr_parts, parts, \
				s_env, &exec);
		}
		i++;
	}
	if (exec == 1)
		return (1);
	return (0);
}

int	check_for_pipes(int *exit_status, t_part *parts, \
t_env *s_env, int nr_parts)
{
	int	i;

	i = 0;
	while (i < nr_parts)
	{
		if (parts[i].type == SPECIAL && ft_strchr(parts[i].part, '|'))
		{
			if (i == nr_parts - 1)
				*exit_status = ft_syntax_error_eof(s_env->line_nr);
			else if (ft_strcmp(parts[i].part, "|") || !i \
			|| ft_strchr(parts[i + 1].part, '|'))
				*exit_status = ft_syntax_error(parts, i, s_env->line_nr);
			else
				*exit_status = ft_pipex(nr_parts, parts, s_env);
			return (1);
		}
		i++;
	}
	return (0);
}

int	ft_check_parts(int *last_exit_status, char *line, t_env *s_env)
{
	t_part	*parts;
	int		nr_parts;
	int		executed;

	parts = ft_shell_split(line, *last_exit_status, s_env);
	nr_parts = count_parts(parts);
	if (nr_parts)
	{
		executed = check_for_pipes(last_exit_status, parts, \
		s_env, nr_parts);
		if (!executed)
			executed = check_for_redirections(last_exit_status, parts, s_env, \
			nr_parts);
		if (!executed)
			executed = is_built_in(parts[0].part, nr_parts, parts, s_env);
		if (executed == 300)
			*last_exit_status = ft_executable(nr_parts, parts, s_env);
	}
	free(line);
	ft_free_parts(parts);
	return (*last_exit_status);
}

int	loop_through_lines(char *line, int last_exit_status, t_env *s_env)
{
	while (line)
	{
		if (isatty(STDIN_FILENO) && ft_strlen(line))
			add_history(line);
		last_exit_status = ft_check_parts(&last_exit_status, line, s_env);
		if (isatty(STDIN_FILENO))
			line = readline(PROMPT);
		else
			line = get_next_line(STDIN_FILENO);
		s_env->line_nr++;
	}
	return (last_exit_status);
}

int	main(int argc, char **argv, char **envp)
{
	char	*line;
	int		last_exit_status;
	t_env	s_env;

	(void)argc;
	(void)argv;
	s_env.line_nr = 1;
	setbuf(stdout, NULL);
	signals();
	copy_env(envp, &s_env);
	if (isatty(STDIN_FILENO))
		line = readline(PROMPT);
	else
	{
		line = get_next_line(STDIN_FILENO);
		line = ft_strtrim_free(&line, "\n");
	}
	last_exit_status = 0;
	last_exit_status = loop_through_lines(line, last_exit_status, &s_env);
	ft_free_strs(s_env.env);
	return (last_exit_status);
}
