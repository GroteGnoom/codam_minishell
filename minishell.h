#ifndef MINISHELL_H
# define MINISHELL_H

enum e_part_type {
	SPACES,
	SINGLE_QUOTED,
	DOUBLE_QUOTED,
	NORMAL,
	SPECIAL
};

typedef struct s_part {
	char				*part;
	enum e_part_type	type;
}	t_part;

typedef struct s_pipe
{
	int		infile;
	int		outfile;
	int		size;
	int		iter;
	int		len;
	int		begin;
	int		end;

	char	**commands;
	char	**paths;

	char	*cmd;
	char	**cmd_flag;
}	t_pipe;

typedef struct s_env
{
	char	**env;
	int		size;
}	t_env;

char	**ft_get_commands(char **argv, int len, t_pipe *pipex);
char	**ft_split_commands(char const *s, char c);
void	ft_close_pipes(t_pipe pipex, int *pipefd);
void	ft_child_process(t_pipe pipex, int *pipefd, char **envp);
int		ft_pipex(int argc, char **argv, char **envp);

void	signals(void);

int		ft_echo(char **args, int nr_args);
int		ft_pwd(void);
int		ft_cd(char **args);
int		ft_env(char **envp);
int		ft_export(char **args, t_env *s_env);
int		ft_exit(char **args, int nr_args);
void	copy_env(char **envp, t_env *s_env);
int		ft_unset(char **args, t_env *s_env);

int		ft_executable(char **args, t_env *s_env);
void	expand_args(char **sp, int last_exit_status);
int		ft_export_print(char **envp);
t_part	*quote_split(char *s);
void	expand_unquoted_args(t_part *parts, int last_exit_status);
char	**parts_to_strings(t_part *parts);

int		ft_redirect_in(char **args, t_env *s_env, int argc);
int		ft_redirect_out(char **args, t_env *s_env, int argc);
int		ft_redirect_out_app(char **args, t_env *s_env, int argc);
void	ft_try_paths(char **paths, char **args, char **envp);
#endif
