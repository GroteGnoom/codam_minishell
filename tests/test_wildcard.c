/* ************************************************************************** */
/*                                                                            */
/*                                                        ::::::::            */
/*   test_wildcard.c                                    :+:    :+:            */
/*                                                     +:+                    */
/*   By: dnoom <marvin@codam.nl>                      +#+                     */
/*                                                   +#+                      */
/*   Created: 2021/12/15 14:35:14 by dnoom         #+#    #+#                 */
/*   Updated: 2021/12/16 15:55:32 by dnoom         ########   odam.nl         */
/*                                                                            */
/* ************************************************************************** */

#include <stdio.h>
#include "../minishell.h"
#include "../Libft/libft.h"

int	main(void)
{
	char	*result;

	result = ft_wildcard(ft_strdup("test_wi*c"));
	printf("%s\n\n", result);
	free(result);
	result = ft_wildcard(ft_strdup("test_wi*.c"));
	printf("%s\n\n", result);
	free(result);
	result = ft_wildcard(ft_strdup("*wildcard.c"));
	printf("%s\n\n", result);
	free(result);
	result = ft_wildcard(ft_strdup("*wild*a*r*.c"));
	printf("%s\n\n", result);
	free(result);
	result = ft_wildcard(ft_strdup("*wild**.c"));
	printf("%s\n\n", result);
	free(result);
}
