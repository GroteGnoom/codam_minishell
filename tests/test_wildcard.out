test_wildcard.c

test_wildcard.c

test_wildcard.c

test_wildcard.c

test_wildcard.c

